src/telegraf.ts

```diff
@@ -211,12 +216,18 @@ export class Telegraf<C extends Context = Context> extends Composer<C> {
     const TelegrafContext = this.options.contextType
     const ctx = new TelegrafContext(update, tg, this.botInfo)
     Object.assign(ctx, this.context)
+    const timeoutsAt = Date.now() + this.options.handlerTimeout
+    const done =
+      this.options.handlerTimeout <= NODEJS_MAX_TIMER_DURATION
+        ? this.timeouts.add({ ctx, timeoutsAt })
+        : anoop
     try {
       await this.middleware()(ctx, anoop)
     } catch (err) {
       return await this.handleError(err, ctx)
     } finally {
-      if (webhookResponse !== undefined && !webhookResponse.writableEnded) {
+      done()
+      if (webhookResponse?.writableEnded === false) {
```

Comment:
> 3. How come you have such a strict separation between different ways how middleware can fail? If it throws, you handle it here, if it times out, you don't. Why distribute handling failure across different modules?

Excellent point. I'm considering changing it to something akin to:

```ts
await this.timeouts.add(this.middleware(), ctx)
```

Promise returned by `.add` would either

- resolve if middleware succeeds,
- reject with `TimeoutError`,
- do whatever `handleError` does in case of error.

This would keep error handling straightforward; no matter what happens, all `finally` handlers will run.

What do you think?

Reply:
Let's merge #1258 if you like it, before I create conflicts.

Reply:
Better idea: remove `handleTimeout` entirely, signal them as passing a `TimeoutError` to `handleError`!

Reply:
Almost ready, one type error to sort out, I can push if you want.

Reply:
Not quite sure how it will look then. Also not sure if it's possible to fix it without refactoring `Timeouts`, hence also not sure if I want to invest more time in this PR

Reply:
Anything I say here is pointing out how #1247 solves things better than this branch. We're reinventing the wheel here, because you don't want to read the ready solution?

Reply:
I genuinely think my code is superior because of its simplicity. #1247 is concerning to me. Tomorrow I'll sit down and review it thoroughly to explain why.